src/lib.rs

```diff
@@ -330,10 +331,11 @@ impl fmt::Display for GlobError {
 struct PathWrapper {
     path: PathBuf,
     is_directory: bool,
+    file_name: Option<OsString>,
```

Comment:
It seems odd to me that an extra allocation for each path speeds things up. This is just to avoid the work of finding the filename in a path, right?

Have you tried storing the indexes into `path` and reconstituting an &OsStr by slicing into path?

Reply:
> This is just to avoid the work of finding the filename in a path, right?

Yes, with a caveat that on the current main version we end up performing a filename search multiple times per each path, as it's done in a sort comparator. That's where the crux of a speedup of this PR comes from.

No, I have not tried to store the subrange of a filename. I'll give it a go, thank you for the suggestion!

Reply:
Ah the necessary methods needed to do the slicing don't seem to be msrv-compatible, a pity.

A possible workaround for that is to do the slicing on `&str` and fall back to pulling the `file_name` out of the pathbuf on every access when it's not utf8-compatible (which should be rare).

Reply:
That'd require either running UTF-8 validation on `PathWrapper::path` whenever we try to use a filename slice (or introducing `unsafe` into the lib to avoid the repeated validation); I wonder if the extra complexity is worth it in the first place.
FWIW, b7dbb48ee893ec685449d117f79c50bae970bbe4 increases peak memory usage of a `rust-lang/rust` benchmark by about 23%. I've managed to reduce this to 6% by replacing `OsString` with `Box<OsStr>` in 778365d; I agree that allocating here is not ideal, but I'm unsure about our options here.

Reply:
I think we can get away without unsafe. We can go Path -> Osstr -> Option<&str> -> Option<&[u8]> and then compare the [u8]s most of the time and only in non-utf8 cases use the old code path.

Reply:
We'd have to do that each time we need to inspect the filename, right? We can't store a reference to `PathWrapper::path` within PathWrapper itself.
I believe that `OsStr` -> `Option<&str>` might entail a UTF-8 validation. It'd be fine to do it once, but if above is correct, we'd have to do it on each filename access.

Reply:
Ah right. I guess we can store a pointer instead of a reference. That'd require a bit of unsafe after all, but should be ok since Pathbuf contents are heap-allocated and don't move.